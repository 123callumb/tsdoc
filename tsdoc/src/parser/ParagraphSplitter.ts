import {
  DocComment,
  DocSection,
  DocNode,
  DocNodeKind,
  DocParagraph,
  DocPlainText
} from '../nodes';

export class ParagraphSplitter {
  private static readonly _whitespaceRegExp: RegExp = /^\s*$/;

  /**
   * For each DocSection, split the DocParagraph into multiple paragraphs by looking
   * for one or more blank lines that separate the paragraphs.  (These "lines" are ended
   * by SoftBreak nodes).  The blank lines are assigned to the preceding paragraph,
   * and referred to as the "trailer".
   */
  public static splitParagraphs(docComment: DocComment): void {
    for (const node of docComment.getChildNodes()) {
      if (node instanceof DocSection) {
        ParagraphSplitter._splitParagraphsForSection(node);
      }
    }
  }

  private static _splitParagraphsForSection(docSection: DocSection): void {
    const inputNodes: ReadonlyArray<DocNode> = docSection.nodes;
    const outputNodes: DocNode[] = [];

    for (const oldNode of inputNodes) {
      if (oldNode.kind === DocNodeKind.Paragraph) {
        ParagraphSplitter._splitParagraph(oldNode as DocParagraph, outputNodes);
      } else {
        outputNodes.push(oldNode);
      }
    }

    // Replace the inputNodes with the outputNodes
    docSection.clearNodes();
    docSection.appendNodes(outputNodes);
  }

  private static _splitParagraph(oldParagraph: DocParagraph, outputNodes: DocNode[]): void {
    const inputParagraphNodes: ReadonlyArray<DocNode> = oldParagraph.nodes;

    let currentParagraph: DocParagraph = new DocParagraph({ });
    outputNodes.push(currentParagraph);

    const enum SplitterState {
      Start,
      AwaitingTrailer,
      ReadingTrailer
    }
    let state: SplitterState = SplitterState.Start;

    let currentIndex: number = 0;
    while (currentIndex < inputParagraphNodes.length) {

      // Scan forwards to the end of the line
      let isBlankLine: boolean = true;
      let lineEndIndex: number = currentIndex; // non-inclusive
      do {
        const node: DocNode = inputParagraphNodes[lineEndIndex++];
        if (node.kind === DocNodeKind.SoftBreak) {
          break;
        }
        if (isBlankLine) {
          if (!this._isWhitespace(node)) {
            isBlankLine = false;
          }
        }
      } while (lineEndIndex < inputParagraphNodes.length);

      // At this point, the line and SoftBreak will be in inputParagraphNodes.slice(currentIndex, lineEndIndex)

      switch (state) {
        case SplitterState.Start:
          // We're skipping any blank lines that start the first paragraph
          if (!isBlankLine) {
            state = SplitterState.AwaitingTrailer;
          }
          break;
        case SplitterState.AwaitingTrailer:
          // We already saw some content, so now we're looking for a blank line that starts the trailer
          // at the end of this paragraph
          if (isBlankLine) {
            state = SplitterState.ReadingTrailer;
          }
          break;
        case SplitterState.ReadingTrailer:
          // We already found the trailer, so now we're looking for a non-blank line that will
          // begin a new paragraph
          if (!isBlankLine) {
            // Start a new paragraph
            currentParagraph = new DocParagraph({ });
            outputNodes.push(currentParagraph);

            state = SplitterState.AwaitingTrailer;
          }
          break;
      }

      // Append the line onto the current paragraph
      for (let i: number = currentIndex; i < lineEndIndex; ++i) {
        currentParagraph.appendNode(inputParagraphNodes[i]);
      }

      currentIndex = lineEndIndex;
    }
  }

  private static _isWhitespace(node: DocNode): boolean {
    switch (node.kind) {
      case DocNodeKind.PlainText:
        const docPlainText: DocPlainText = node as DocPlainText;
        return ParagraphSplitter._whitespaceRegExp.test(docPlainText.text);
      default:
        return false;
    }
  }
}
