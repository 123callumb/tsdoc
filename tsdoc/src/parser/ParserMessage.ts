import { TextRange, ITextLocation } from './TextRange';
import { TokenSequence } from './TokenSequence';
import { DocNode } from '../nodes/DocNode';

/**
 * Constructor parameters for {@link ParserMessage}.
 */
export interface IParserMessageParameters {
  messageText: string;
  textRange: TextRange;
  tokenSequence?: TokenSequence;
  docNode?: DocNode;
}

/**
 * Represents an error or warning that occurred during parsing.
 */
export class ParserMessage {
  /**
   * The message text.
   */
  public readonly text: string;

  /**
   * The message text without the default prefix that shows line/column information.
   */
  public readonly unformattedText: string;

  public readonly textRange: TextRange;

  public readonly tokenSequence: TokenSequence | undefined;

  public readonly docNode: DocNode | undefined;

  /**
   * Generates a line/column prefix.  Example with line=2 and column=5
   * and message="An error occurred":
   * ```
   * "(2,5): An error occurred"
   * ```
   */
  private static _formatMessageText(message: string, range: TextRange): string {
    if (!message) {
      message = 'An unknown error occurred';
    }

    if (range.pos !== 0 || range.end !== 0) {
      const location: ITextLocation = range.getLocation(range.pos);
      if (location.line) {
        return `(${location.line},${location.column}): ` + message;
      }
    }
    return message;
  }

  public constructor(parameters: IParserMessageParameters) {
    this.unformattedText = parameters.messageText;
    this.textRange = parameters.textRange;
    this.tokenSequence = parameters.tokenSequence;
    this.docNode = parameters.docNode;

    this.text = ParserMessage._formatMessageText(this.unformattedText, this.textRange);
  }

  public toString(): string {
    return this.text;
  }
}
