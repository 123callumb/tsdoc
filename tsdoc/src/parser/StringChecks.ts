/**
 * Helpers for validating various text string formats.
 */
export class StringChecks {
  private static readonly tsdocTagNameRegExp: RegExp = /^@[a-z][a-z0-9]*$/i;

  /**
   * Tests tagName to see whether it is a valid TSDoc tag name; if not, returns an error message.
   * TSDoc tag names start with an at-sign ("@") followed by ASCII letters using
   * "camelCase" capitalization.
   */
  public static explainIfNotTSDocTagName(tagName: string): string | undefined {
    if (tagName[0] !== '@') {
      return 'A TSDoc tag name must start with an "@" symbol';
    }

    if (!StringChecks.tsdocTagNameRegExp.test(tagName)) {
      return 'A TSDoc tag name must start with a letter and contain only letters and numbers';
    }

    return undefined;
  }

  /**
   * Throws an exception if the input is not a valid TSDoc tag name.
   * TSDoc tag names start with an at-sign ("@") followed by ASCII letters using
   * "camelCase" capitalization.
   */
  public static validateTSDocTagName(tagName: string): void {
    const explanation: string | undefined = StringChecks.explainIfNotTSDocTagName(tagName);
    if (explanation) {
      throw new Error(explanation);
    }
  }
}
