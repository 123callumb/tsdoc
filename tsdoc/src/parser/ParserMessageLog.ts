import { ParserMessage } from './ParserMessage';
import { TextRange } from './TextRange';
import { TokenSequence } from './TokenSequence';
import { DocNode } from '../nodes/DocNode';

/**
 * Used to report errors and warnings that occurred during parsing.
 */
export class ParserMessageLog {
  private _messages: ParserMessage[] = [];

  /**
   * The unfiltered list of all messages.
   */
  public get messages(): ReadonlyArray<ParserMessage> {
    return this._messages;
  }

  /**
   * Append a message to the log.
   */
  public addMessage(parserMessage: ParserMessage): void {
    this._messages.push(parserMessage);
  }

  /**
   * Append a message associated with a TextRange.
   */
  public addMessageForTextRange(messageText: string, textRange: TextRange): void {
    this.addMessage(new ParserMessage({
      messageText,
      textRange
    }));
  }

  /**
   * Append a message associated with a TokenSequence.
   */
  public addMessageForTokenSequence(messageText: string, tokenSequence: TokenSequence, docNode?: DocNode): void {
    this.addMessage(new ParserMessage({
      messageText,
      textRange: tokenSequence.parserContext.tokens[tokenSequence.startIndex].range,
      tokenSequence,
      docNode
    }));
  }
}
