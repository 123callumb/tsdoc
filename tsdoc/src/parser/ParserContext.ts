import { TextRange } from './TextRange';
import { Token } from './Token';
import { DocComment, DocNode } from '../nodes';
import { TSDocParserConfiguration } from './TSDocParserConfiguration';
import { ParserMessageLog } from './ParserMessageLog';

/**
 * An internal data structure that tracks all the state being built up by the various
 * parser stages.
 */
export class ParserContext {
  /**
   * The configuration that was provided for the TSDocParser.
   */
  public readonly configuration: TSDocParserConfiguration;

  /**
   * The `sourceRange` indicates the start and end of the original input that was parsed.
   */
  public readonly sourceRange: TextRange;

  /**
   * The text range starting from the opening `/**` and ending with
   * the closing `*\/` delimiter.
   */
  public commentRange: TextRange = TextRange.empty;

  /**
   * The text ranges corresponding to the lines of content inside the comment.
   */
  public lines: TextRange[] = [];

  /**
   * A complete list of all tokens that were extracted from the input lines.
   */
  public tokens: Token[] = [];

  /**
   * A diagnostic list of all DocNode objects that were encountered during the first pass
   * of the parser, in the order that they originally appeared in the input, with no gaps
   * or overlap of token sequences.
   *
   * This linearity may be lost when the nodes are assembled into a DocComment, due to:
   * - reordering/grouping of sections
   * - deleting block tags that were moved to the ModifierTagSet
   * - pruning of whitespace
   */
  public verbatimNodes: DocNode[];

  /**
   * The parsed doc comment object.  This is the primary output of the parser.
   */
  public readonly docComment: DocComment;

  /**
   * A queryable log that reports warnings and error messages that occurred during parsing.
   */
  public readonly log: ParserMessageLog;

  public constructor(configuration: TSDocParserConfiguration, sourceRange: TextRange) {
    this.configuration = configuration;
    this.sourceRange = sourceRange;

    this.verbatimNodes = [];

    this.docComment = new DocComment({ parserContext: this });

    this.log = new ParserMessageLog();
  }
}
