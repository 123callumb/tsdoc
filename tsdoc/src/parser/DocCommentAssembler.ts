import { ParserContext } from './ParserContext';
import {
  DocNodeKind,
  DocBlockTag,
  DocComment,
  DocNode
} from '../nodes';
import {
  TSDocTagDefinition,
  TSDocParserConfiguration,
  TSDocTagSyntaxKind
} from './TSDocParserConfiguration';
import { ModifierTagSet } from '../details/ModifierTagSet';

/**
 * After the NodeParser has constructed the  ParserContext.verbatimSection,
 * the DocCommentAssembler reorganizes these nodes into a DocComment object.
 */
export class DocCommentAssembler {
  private _configuration: TSDocParserConfiguration;
  private _parserContext: ParserContext;
  private _docComment: DocComment;

  public constructor(parserContext: ParserContext) {
    this._parserContext = parserContext;
    this._configuration = parserContext.configuration;
    this._docComment = parserContext.docComment;
  }

  public assemble(): void {

    const modifierTagSet: ModifierTagSet = this._docComment.modifierTagSet;

    // Scan all the top-level nodes and weed out the modifier tags
    const prunedDocNodes: DocNode[] = [];

    for (const docNode of this._parserContext.verbatimSection.getChildNodes()) {

      switch (docNode.kind) {
        case DocNodeKind.BlockTag:
          const docBlockTag: DocBlockTag = docNode as DocBlockTag;
          // Do we have a definition for this tag?
          const tagDefinition: TSDocTagDefinition | undefined
            = this._configuration.tryGetTagDefinitionUpperCase(docBlockTag.tagNameForComparisons);
          if (tagDefinition) {
            if (tagDefinition.syntaxKind === TSDocTagSyntaxKind.ModifierTag) {
              modifierTagSet.addModifierTag(docBlockTag);
            }
          }
          break;
        default:
          prunedDocNodes.push(docNode);
          break;
      }
    }

    this._docComment.remarks.appendNodes(prunedDocNodes);
  }
}
