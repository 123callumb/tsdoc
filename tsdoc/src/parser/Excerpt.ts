import { ParserContext } from './ParserContext';
import { TokenSequence } from './TokenSequence';

export interface IExcerptParameters {
  prefix: TokenSequence;
  suffix?: TokenSequence;
  separator?: TokenSequence;
}

/**
 * When a DocNode is constructed by parsing input text, the Excerpt object is used to
 * annotate each node with the associated tokens that were parsed.  This is useful
 * e.g. for highlighting the corresponding input characters to indicate a refactoring
 * match or error location.
 *
 * @remarks
 * The excerpt is separated into three token sequences: the "prefix" (which is the main textual
 * content for the node), the "suffix" (which captures a closing delimiter which sometimes
 * appears after child nodes), and the "separator" which captures whitespace separating
 * child nodes.
 *
 * For example, suppose the TSDoc parser processes this input: `<a href="#" target="_top">`
 * The parent node will be DocHtmlStartTag with prefix=`<a ` and suffix=`>` with no separator.
 * There will be two child nodes of type DocHtmlAttribute.  The first one will have
 * prefix=`href="#"`, no suffix (since it has no children) and separator=` `.
 */
export class Excerpt {
  /**
   * The ParserContext object from the parser invocation that processed the input.
   */
  public readonly parserContext: ParserContext;

  /**
   * The main textual content for the associated node.
   */
  public readonly prefix: TokenSequence;

  /**
   * An optional suffix that captures any delimiters that might appear after the
   * child nodes.  If there are no child nodes, this is always an empty list.
   */
  public readonly suffix: TokenSequence;

  /**
   * Captures any whitespace that may separate this node from a sibling that follows it.
   * The tokens will always be of type Spacing or Newline.
   */
  public readonly separator: TokenSequence;

  public constructor(parameters: IExcerptParameters) {
    this.parserContext = parameters.prefix.parserContext;
    this.prefix = parameters.prefix;
    this.suffix = parameters.suffix || TokenSequence.createEmpty(this.parserContext);
    this.separator = parameters.separator || TokenSequence.createEmpty(this.parserContext);
  }
}
