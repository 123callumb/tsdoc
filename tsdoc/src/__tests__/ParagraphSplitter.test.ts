import { TestHelpers } from '../parser/__tests__/TestHelpers';
import { ParagraphSplitter } from '../parser/ParagraphSplitter';
import { DocSection, DocPlainText, DocSoftBreak, DocParagraph, DocBlockTag } from '../index';

test('01 Basic paragraph splitting', () => {
  TestHelpers.parseAndMatchDocCommentSnapshot([
    '/**',
    ' *    ',
    ' * This is the',
    ' * first paragraph.',
    ' *   \t   ',
    ' *  ',
    ' *   \t   ',
    ' * This is the second paragraph.',
    ' *',
    ' * This is the third paragraph.',
    ' *',
    ' *   ',
    ' */'
  ].join('\n'));
});

test('02 Degenerate paragraph', () => {
  TestHelpers.parseAndMatchDocCommentSnapshot([
    '/** line 1',
    ' * line 2',
    '',
    ' *   @public line 3*/'
  ].join('\n'));
});

test('03 Degenerate manually constructed nodes', () => {
  const docSection: DocSection = new DocSection({ });

  const docParagraph: DocParagraph = new DocParagraph({ } );
  docParagraph.appendNodes([
    new DocPlainText({ text: '  para 1 ' }),
    new DocSoftBreak({ }),
    new DocPlainText({ text: '   ' }),
    new DocSoftBreak({ }),
    new DocPlainText({ text: ' \t  ' }),
    new DocPlainText({ text: '   ' }),
    new DocBlockTag({ tagName: '@public' }),
    new DocPlainText({ text: '  para 2 ' }),
    new DocSoftBreak({ }),
    new DocSoftBreak({ }),
    new DocPlainText({ text: '  para 3  ' })
  ]);

  docSection.appendNode(docParagraph);

  // Currently we do not discard empty paragraphs
  docSection.appendNode(new DocParagraph({ }));

  ParagraphSplitter.splitParagraphsForSection(docSection);
  expect(TestHelpers.getDocNodeSnapshot(docSection)).toMatchSnapshot();
});
