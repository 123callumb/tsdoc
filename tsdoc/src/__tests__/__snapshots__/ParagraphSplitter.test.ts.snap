// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`01 Basic paragraph splitting 1`] = `
Object {
  "_0_lines": Array [
    "",
    "This is the",
    "first paragraph.",
    "",
    "",
    "",
    "This is the second paragraph.",
    "",
    "This is the third paragraph.",
    "",
    "",
  ],
  "_1_gaps": Array [],
  "_2_summarySection": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "This is the",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "first paragraph.",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "This is the second paragraph.",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "This is the third paragraph.",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_3_remarksBlock": undefined,
  "_4_customBlocks": Array [],
  "_5_paramBlocks": Array [],
  "_6_returnsBlock": undefined,
  "_7_modifierTags": Array [],
  "_8_logMessages": Array [],
}
`;

exports[`02 Degenerate paragraph 1`] = `
Object {
  "_0_lines": Array [
    "line 1",
    "line 2",
    "",
    "  @public line 3",
  ],
  "_1_gaps": Array [],
  "_2_summarySection": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "line 1",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "line 2",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "  ",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " line 3",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_3_remarksBlock": undefined,
  "_4_customBlocks": Array [],
  "_5_paramBlocks": Array [],
  "_6_returnsBlock": undefined,
  "_7_modifierTags": Array [
    Object {
      "kind": "BlockTag",
      "nodeExcerpt": "@public",
    },
  ],
  "_8_logMessages": Array [],
}
`;

exports[`03 Degenerate manually constructed nodes 1`] = `
Object {
  "kind": "Section",
  "nodes": Array [
    Object {
      "kind": "Paragraph",
      "nodes": Array [
        Object {
          "kind": "PlainText",
          "nodePlainText": "  para 1 ",
        },
        Object {
          "kind": "SoftBreak",
        },
        Object {
          "kind": "PlainText",
          "nodePlainText": "   ",
        },
        Object {
          "kind": "SoftBreak",
        },
      ],
    },
    Object {
      "kind": "Paragraph",
      "nodes": Array [
        Object {
          "kind": "PlainText",
          "nodePlainText": " [t]  ",
        },
        Object {
          "kind": "PlainText",
          "nodePlainText": "   ",
        },
        Object {
          "kind": "BlockTag",
        },
        Object {
          "kind": "PlainText",
          "nodePlainText": "  para 2 ",
        },
        Object {
          "kind": "SoftBreak",
        },
        Object {
          "kind": "SoftBreak",
        },
      ],
    },
    Object {
      "kind": "Paragraph",
      "nodes": Array [
        Object {
          "kind": "PlainText",
          "nodePlainText": "  para 3  ",
        },
      ],
    },
    Object {
      "kind": "Paragraph",
    },
  ],
}
`;
